### PR TITLE
Effects support for summons

### PR DESCRIPTION
**Pull request by reyaleman**
### Pull Request Prelude

- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
- Added effect support for summons
- Fixed? summon effect to match cipbia (teleport effect on summons not master)
- Added customizable effect for summons (effect)
- Added customizable effect for summon master (masterEffect)
- Added summon structure fields to lua methods (monsterType:getSummonList & monsterType:addSummon)

```xml
<summon name="fire elemental" interval="2000" chance="100">
	<attribute key="mastereffect" value="redteleport" />
	<attribute key="effect" value="hearts" />
</summon>
```

https://user-images.githubusercontent.com/3027705/170199886-7a889e9d-2f9b-4acc-aa47-79aa877b55ba.mp4

**Issues addressed:** 

#4136

[cont]: https://github.com/otland/forgottenserver/wiki/Contributing
[code]: https://github.com/otland/forgottenserver/wiki/TFS-Coding-Style-Guide

### CODE REVIEW

File: src/luascript.cpp [hunk 1]
```diff
@@ -14519,25 +14519,30 @@ int LuaScriptInterface::luaMonsterTypeGetSummonList(lua_State* L)
...
+		setField(L, "max", summonBlock.max);
 		lua_rawseti(L, -2, ++index);
 	}
 	return 1;
 }
 
 int LuaScriptInterface::luaMonsterTypeAddSummon(lua_State* L)
 {
-	// monsterType:addSummon(name, interval, chance[, max = -1])
+	// monsterType:addSummon(name, interval, chance[, effect = CONST_ME_TELEPORT[, masterEffect = CONST_ME_NONE[, max = -1]]])
```
Comment (Codinablack) [inline src/luascript.cpp:14536]: Shouldn't the two new parameters go to the end, to maintain backwards compatibility?

Comment (nekiro) [reply src/luascript.cpp:14536]: it should, good call

File: src/monsters.h [hunk 2]
```diff
@@ -54,6 +54,8 @@ struct summonBlock_t
 	uint32_t speed;
 	uint32_t max;
 	bool force = false;
+	MagicEffectClasses masterEffect = CONST_ME_NONE;
+	MagicEffectClasses effect = CONST_ME_TELEPORT;
```
Comment (ramon-bernardo) [inline src/monsters.h:58]:
```suggestion
	MagicEffectClasses effect = CONST_ME_TELEPORT;
	MagicEffectClasses masterEffect = CONST_ME_NONE;
	bool force = false;
```

File: src/monsters.cpp [hunk 3]
```diff
@@ -1347,13 +1349,44 @@ MonsterType* Monsters::loadMonster(const std::string& file, const std::string& m
...
+
 			if ((attr = summonNode.attribute("name"))) {
 				summonBlock_t sb;
 				sb.name = attr.as_string();
 				sb.speed = speed;
 				sb.chance = chance;
 				sb.max = max;
 				sb.force = force;
+				sb.masterEffect = masterEffect;
+				sb.effect = effect;
```
Comment (ramon-bernardo) [inline src/monsters.cpp:1389]:
```suggestion
				sb.effect = effect;
				sb.masterEffect = masterEffect;
				sb.force = force;
```

File: src/monsters.cpp [hunk 4]
```diff
@@ -1347,13 +1349,44 @@ MonsterType* Monsters::loadMonster(const std::string& file, const std::string& m
 				force = attr.as_bool();
 			}
 
+			for (const auto& attributeNode : summonNode.children()) {
```
Comment (ramon-bernardo) [inline src/monsters.cpp:1352]: To above "force" attr

File: src/monsters.cpp [hunk 5]
```diff
@@ -1325,6 +1325,8 @@ MonsterType* Monsters::loadMonster(const std::string& file, const std::string& m
 			int32_t speed = 1000;
 			int32_t max = mType->info.maxSummons;
 			bool force = false;
+			MagicEffectClasses masterEffect = CONST_ME_NONE;
+			MagicEffectClasses effect = CONST_ME_TELEPORT;
```
Comment (ramon-bernardo) [inline src/monsters.cpp:1329]:
```suggestion
			MagicEffectClasses effect = CONST_ME_TELEPORT;
			MagicEffectClasses masterEffect = CONST_ME_NONE;
			bool force = false;
```

**Review by ramon-bernardo — COMMENTED** [review]
LGTM

Some suggestions

**Review by ranisalt — APPROVED** [review]

**Review by EPuncker — APPROVED** [review]

**Comment by EPuncker** [comment]
LGTM

File: src/monsters.h [hunk 6]
```diff
@@ -56,6 +56,8 @@ struct summonBlock_t
 	uint32_t chance;
 	uint32_t speed;
 	uint32_t max;
+	MagicEffectClasses effect = CONST_ME_TELEPORT;
+	MagicEffectClasses masterEffect = CONST_ME_NONE;
```
Comment (ramon-bernardo) [inline src/monsters.h:60]: No default value required, since everywhere we create summonBlock_t, it is overridden

**Review by ramon-bernardo — APPROVED** [review]

**Pull request merged**